Content.Server/Anomaly/Effects/FleshAnomalySystem.cs

```diff
@@ -0,0 +1,102 @@
+    }
+
+    private void OnPulse(EntityUid uid, FleshAnomalyComponent component, ref AnomalyPulseEvent args)
+    {
+        var range = component.SpawnRange * args.Stability;
+        var amount = (int) (component.MaxSpawnAmount * args.Severity + 0.5f);
+
+        var xform = Transform(uid);
+        SpawnMonstersOnOpenTiles(component, xform, amount, range);
+    }
+
+    private void OnSupercritical(EntityUid uid, FleshAnomalyComponent component, ref AnomalySupercriticalEvent args)
+    {
+        var xform = Transform(uid);
+        SpawnMonstersOnOpenTiles(component, xform, component.MaxSpawnAmount, component.SpawnRange);
+        Spawn(component.SupercriticalSpawn, xform.Coordinates);
+    }
+
+    private void OnSeverityChanged(EntityUid uid, FleshAnomalyComponent component, ref AnomalyStabilityChangedEvent args)
+    {
+        var xform = Transform(uid);
+        if (!_map.TryGetGrid(xform.GridUid, out var grid))
+            return;
+
+        var radius = component.SpawnRange * args.Stability;
+        var fleshTile = (ContentTileDefinition) _tiledef[component.FleshTileId];
+        var localpos = xform.Coordinates.Position;
+        var tilerefs = grid.GetLocalTilesIntersecting(
+            new Box2(localpos + (-radius, -radius), localpos + (radius, radius)));
+        foreach (var tileref in tilerefs)
+        {
+            if (!_random.Prob(0.33f))
+                continue;
+            _tile.ReplaceTile(tileref, fleshTile);
+        }
+    }
+
+    private void SpawnMonstersOnOpenTiles(FleshAnomalyComponent component, TransformComponent xform, int amount, float radius)
+    {
+        if (!_map.TryGetGrid(xform.GridUid, out var grid))
+            return;
+
+        var localpos = xform.Coordinates.Position;
+        var tilerefs = grid.GetLocalTilesIntersecting(
+            new Box2(localpos + (-radius, -radius), localpos + (radius, radius)));
+        var validSpawnLocations = new List<Vector2i>();
+        var physQuery = GetEntityQuery<PhysicsComponent>();
+        foreach (var tileref in tilerefs)
+        {
+            var valid = true;
+            foreach (var ent in grid.GetAnchoredEntities(tileref.GridIndices))
+            {
+                if (!physQuery.TryGetComponent(ent, out var body))
+                    continue;
+                if (body.BodyType == BodyType.Static &&
+                    body.Hard &&
+                    (body.CollisionLayer & (int) CollisionGroup.Impassable) != 0)
+                {
+                    valid = false;
+                    break;
+                }
+            }
+            if (valid)
+                validSpawnLocations.Add(tileref.GridIndices);
+        }
+
+        if (!validSpawnLocations.Any())
+            return;
+
+        for (var i = 0; i < amount; i++)
+        {
+            var pos = _random.Pick(validSpawnLocations);
+            Spawn(_random.Pick(component.Spawns), pos.ToEntityCoordinates(xform.GridUid.Value, _map));
+        }
```

Comment:
can you make this more compact if you spawn inside the `tileref` loop and shuffle the list instead of picking? (also is it possible for this list to be empty without breaking the loop? should make sure that check is necessary)